docarray/doc_index/backends/hnswlib_doc_index.py

```diff
@@ -0,0 +1,400 @@
+    BaseDocumentIndex,
+    FindResultBatched,
+    _Column,
+)
+from docarray.proto import DocumentProto
+from docarray.utils.filter import filter as da_filter
+from docarray.utils.find import FindResult
+from docarray.utils.misc import is_np_int, torch_imported
+
+if TYPE_CHECKING:
+
+    def composable(fn):  # type: ignore
+        return fn
+
+else:
+    # static type checkers do not like callable objects as decorators
+    from docarray.doc_index.abstract_doc_index import composable
+
+TSchema = TypeVar('TSchema', bound=BaseDocument)
+T = TypeVar('T', bound='HnswDocumentIndex')
+
+HNSWLIB_PY_VEC_TYPES = [list, tuple, np.ndarray]
+if torch_imported:
+    import torch
+
+    HNSWLIB_PY_VEC_TYPES.append(torch.Tensor)
+
+
+class HnswDocumentIndex(BaseDocumentIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._db_config = cast(HnswDocumentIndex.DBConfig, self._db_config)
+        self._work_dir = self._db_config.work_dir
+        load_existing = os.path.exists(self._work_dir) and os.listdir(self._work_dir)
+        Path(self._work_dir).mkdir(parents=True, exist_ok=True)
+
+        # HNSWLib setup
+        self._index_construct_params = ('space', 'dim')
+        self._index_init_params = (
+            'max_elements',
+            'ef_construction',
+            'M',
+            'allow_replace_deleted',
+        )
+
+        self._hnsw_locations = {
+            col_name: os.path.join(self._work_dir, f'{col_name}.bin')
+            for col_name, col in self._columns.items()
+            if col.config
+        }
+        self._hnsw_indices = {}
+        for col_name, col in self._columns.items():
+            if not col.config:
+                continue  # do not create column index if no config is given
+            if load_existing:
+                self._hnsw_indices[col_name] = self._load_index(col_name, col)
+            else:
+                self._hnsw_indices[col_name] = self._create_index(col)
+
+        # SQLite setup
+        self._sqlite_db_path = os.path.join(self._work_dir, 'docs_sqlite.db')
+        self._sqlite_conn = sqlite3.connect(self._sqlite_db_path)
+        self._sqlite_cursor = self._sqlite_conn.cursor()
+        self._create_docs_table()
+        self._sqlite_conn.commit()
+
+    ###############################################
+    # Inner classes for query builder and configs #
+    ###############################################
+    class QueryBuilder(BaseDocumentIndex.QueryBuilder):
+        def build(self, *args, **kwargs) -> Any:
+            return self._queries
+
+    @dataclass
+    class DBConfig(BaseDocumentIndex.DBConfig):
+        work_dir: str = '.'
+
+    @dataclass
+    class RuntimeConfig(BaseDocumentIndex.RuntimeConfig):
+        default_column_config: Dict[Type, Dict[str, Any]] = field(
+            default_factory=lambda: {
+                np.ndarray: {
+                    'dim': 128,
```

Comment:
do we need to keep this `dim` here when class `_Column` has `n_dim`?

Reply:
good point, let me double check this

Reply:
Ok to clarify: The `n_dim` in the `_Column` is taken from the type parameter, e.g. `NdArray[512]`, then `n_dim` will be 512. The `dim` here is just a parameter that people can pass to `Field()`. So in the `_Column` it could be that `n_dim` is empty while `dim` isn't, or vice versa, or both are empty, etc.